Enhanced utori-like spells & soul fire rune

raymondtfr opened the pull request:
All the spells have been researched in the most recent test server.
The formulas for **rounds**, **time** _(if not constant)_, and the **damage** _(if not constant)_  are very close to vanilla ones.

- [x] **Soul Fire rune**
- [x] **Holy Flash** *"utori san"*
- [x] **Electrify** *"utori vis"*
- [x] **Ignite** *"utori flam"*
- [x] **Envenom** *"utori pox"*
- [x] **Inflict Wound** *"utori kor"*
- [x] **Curse** *"utori mort"*

All codes have now been moved from ``onCastSpell`` to ``onTargetCreature`` callback function (which fixed the #1625 and is also perfomance-wise.)
Some remaining issues have been solved in #1825 with the help of other contributors.

marksamman commented:
The random value is evaluated when the script is loaded, which I don't think is intended. I'm not sure if it's fixable without leaking memory with the current way conditions work though.

raymondtfr commented:
Oh, thanks for the heads-up. I guess you meant to explain this in #1625 .

marksamman commented:
Yes, we should address that first.

djarek commented:
#1645 provides facilities for creating spells with "dynamic" conditions.

ranisalt commented on data/spells/scripts/attack/envenom.lua:
```diff
@@ -3,10 +3,69 @@ combat:setParameter(COMBAT_PARAM_TYPE, COMBAT_EARTHDAMAGE)
+	[4] = {4, 5},
+	[3] = {6, 7},
+	[2] = {10, 10},
+	[1] = {19, 20}
```
Why don't just write backwards? Also it looks very much like a function, I'll try to find it.

raymondtfr replied:
Because I chose to follow the same "descendant order" pattern as with the damage values. 

Yes, those rounds values are constant, and they are used in conditions (such as Bleeding and Poison) in vanilla, so they might have achieved these values through a function (or not), although IMO it is better written as a table and with the respective values for each damage, as users can easily modify them.

I initially thought about writing a simple function for it though, so we could load them from **(global.lua ?)**.

ranisalt replied:
Writing a function wouldn't prevent one to modify the output. You can easily modify the function to just `return 7` if you want. I think it gets cleaner, especially if other functions use the same table.

raymondtfr replied:
In this case, don't you think it would be better if I'd add this table to spells.lua?

ranisalt replied:
Surely. What I meant initially is that the values look very much something like `dmg(x) = x/20` (check the outputs), not that it should be its own function :stuck_out_tongue_closed_eyes:

raymondtfr replied:
I thinking about doing it like this (in **lib/spells.lua**):
```lua
CONDITION = {
	poison = {
		[20] = {1, 1}, [19] = {1, 2}, [18] = {1, 2}, [17] = {1, 2},   [16] = {1, 2},
		[15] = {1, 2}, [14] = {1, 2}, [13] = {1, 2}, [12] = {1, 2},   [11] = {1, 2},
		[10] = {1, 2}, [9]  = {2, 3}, [8]  = {2, 3}, [7]  = {2, 3},   [6]  = {3, 4},
		[5]  = {3, 4}, [4]  = {4, 5}, [3]  = {6, 7}, [2]  = {10, 10}, [1]  = {19, 20}
	},

	bleeding = {
		[20] = {1, 1}, [19] = {1, 2}, [18] = {1, 2}, [17] = {1, 2},   [16] = {1, 2},
		[15] = {1, 2}, [14] = {1, 2}, [13] = {1, 2}, [12] = {1, 2},   [11] = {1, 2},
		[10] = {1, 2}, [9]  = {2, 3}, [8]  = {2, 3}, [7]  = {2, 3},   [6]  = {3, 4},
		[5]  = {3, 4}, [4]  = {4, 5}, [3]  = {6, 7}, [2]  = {10, 10}, [1]  = {19, 20}
	},

	cursed = {
		{rounds = 3, 2},
		{rounds = 2, 3},
		{rounds = 1, 4, 5, 6, 7, 8, 10, 12, 14, 17, 21, 25, 30, 36, 43, 52, 62, 74, 89, 107, 128, 154, 185, 222, 266, 319, 383, 460, 552, 662, 795, 954}
	}
}

In addition to this, I would add more values in Curse table.
```

raymondtfr commented:
Because I chose to follow the same "descendant order" pattern as with the damage values. 

Yes, those rounds values are constant, and they are used in conditions (such as Bleeding and Poison) in vanilla, so they might have achieved these values through a function (or not), although IMO it is better written as a table and with the respective values for each damage, as users can easily modify them.

I initially thought about writing a simple function for it though, so we could load them from **(global.lua ?)**.

raymondtfr commented:
@marksamman After so many changes seeking for improvements, I finally believe this PR is ready (as stated in my first post). Can you review this PR?

marksamman commented:
Looks good to me, but it will have to wait until after the 1.2 release.

raymondtfr commented:
I have now removed curse condition table from ``lib/spells.lua`` and instead I added a function (with shorten and improved code) which generates a damage list for the condition and adds them. The type of generated damage list can be specified in the first argument of the function. I have decided to use a number type (not a condition type) for the user to choose which damage list to generate, since it would be limiting if I'd chosen to check the condition type or similar.

marksamman commented:
Thanks!
